services/src/main/java/org/keycloak/credential/WebAuthnCredentialProvider.java

```diff
@@ -0,0 +1,257 @@
+        }
+        // delete webauthn authenticator's metadata
+        user.removeAttribute(WebAuthnConstants.PUBKEY_CRED_AAGUID_ATTR);
+        user.removeAttribute(WebAuthnConstants.PUBKEY_CRED_ID_ATTR);
+        user.removeAttribute(WebAuthnConstants.PUBKEY_CRED_LABEL_ATTR);
+    }
+
+    @Override
+    public Set<String> getDisableableCredentialTypes(RealmModel realm, UserModel user) {
+        return isConfiguredFor(realm, user, WebAuthnCredentialModel.WEBAUTHN_CREDENTIAL_TYPE) ? Collections.singleton(WebAuthnCredentialModel.WEBAUTHN_CREDENTIAL_TYPE) : Collections.emptySet();
+    }
+
+    @Override
+    public boolean supportsCredentialType(String credentialType) {
+        return WebAuthnCredentialModel.WEBAUTHN_CREDENTIAL_TYPE.equals(credentialType);
+    }
+
+    @Override
+    public boolean isConfiguredFor(RealmModel realm, UserModel user, String credentialType) {
+        if (!supportsCredentialType(credentialType)) return false;
+        return !session.userCredentialManager().getStoredCredentialsByType(realm, user, credentialType).isEmpty();
+    }
+
+    @Override
+    public boolean isValid(RealmModel realm, UserModel user, CredentialInput input) {
+        if (!WebAuthnCredentialModel.class.isInstance(input)) return false;
+
+        WebAuthnCredentialModel context = WebAuthnCredentialModel.class.cast(input);
+        List<WebAuthnCredentialModel> auths = getWebAuthnCredentialModelList(realm, user);
+
+        WebAuthnAuthenticationContextValidator webAuthnAuthenticationContextValidator =
+                new WebAuthnAuthenticationContextValidator();
+        try {
+            for (WebAuthnCredentialModel auth : auths) {
+
+                byte[] credentialId = auth.getAttestedCredentialData().getCredentialId();
+                if (Arrays.equals(credentialId, context.getAuthenticationContext().getCredentialId())) {
+                    Authenticator authenticator = new AuthenticatorImpl(
+                            auth.getAttestedCredentialData(),
+                            auth.getAttestationStatement(),
+                            auth.getCount()
+                    );
+
+                    // WebAuthnException is thrown if validation fails
+                    WebAuthnAuthenticationContextValidationResponse response =
+                            webAuthnAuthenticationContextValidator.validate(
+                                    context.getAuthenticationContext(),
+                                    authenticator);
+
+                    logger.infov("response.getAuthenticatorData().getFlags() = {0}", response.getAuthenticatorData().getFlags());
+
+                    // update authenticator counter
+                    long count = auth.getCount();
+                    auth.setCount(count + 1);
+                    CredentialModel cred = createCredentialModel(auth);
+                    session.userCredentialManager().updateCredential(realm, user, cred);
```

Comment:
I think the counter need to be updated every time the authentication is successfully carried out. The reason is described in https://www.w3.org/TR/webauthn/#signature-counter and https://www.w3.org/TR/webauthn/#verifying-assertion 17 .

Therefore, I would like not to modify this code, but I agree with your idea creating follow-up JIRA to improve the performance later on.

Reply:
I see, thanks for the explain. This raises few interesting questions:
- Firstly in 6.1.1, it is mentioned that "Authenticators SHOULD implement a signature counter feature." . So implemeting it is not mandatory for authenticators per my understanding. I wonder that when authenticator doesn't support counter and will still return value 0, but credential stored on Keycloak side will have value greater than 0, what will happen during validation? Won't the WebAuthnException be thrown in that case? If yes, I am not sure if it is good behaviour as we should take into account that some authenticators may not support the counters?

- Another point - during validation it is mentioned that:
```
Relying Parties should incorporate this information into their risk scoring. Whether the Relying Party updates the stored signature counter value in this case, or not, or fails the authentication ceremony or not, is Relying Party-specific.
```
I can imagine that some admins may prefer stronger performance and avoid saving info during each credential, so maybe we can add the switch whether to support it or not?

Anyway, +1 to do this later. No need to change anything in this PR. Just wanted to clarify this :)

Reply:
>Firstly in 6.1.1, it is mentioned that "Authenticators SHOULD implement a signature counter feature." . So implemeting it is not mandatory for authenticators per my understanding. I wonder that when authenticator doesn't support counter and will still return value 0, but credential stored on Keycloak side will have value greater than 0, what will happen during validation? Won't the WebAuthnException be thrown in that case? If yes, I am not sure if it is good behaviour as we should take into account that some authenticators may not support the counters?

It does not happen that keycloak throws WebAuthnException in such the situation. Webauthn4j implemented WebAuthn specification rigorously. 
https://github.com/webauthn4j/webauthn4j/blob/a166a383b038fc2edbb9859f9540cfae1db8ce2a/webauthn4j-core/src/main/java/com/webauthn4j/validator/WebAuthnAuthenticationContextValidator.java#L170

Reply:
I agree on your idea of adding the switch to control the verification about the counter value in order to avoid frequent write access to DB considering performance. However, I think it can be realized as the other follow-up JIRA. What do you think about it?

Reply:
+1 for adding later. Thanks for pointing the webauthn4j code related to validation.